File: scripts/patchers/kernel_jb_patch_task_conversion.py
"""Mixin: KernelJBPatchTaskConversionMixin."""

import os

from .kernel_jb_base import (
    ARM64_OP_REG,
    ARM64_OP_MEM,
    ARM64_REG_X0,
    ARM64_REG_X1,
    ARM64_REG_W0,
    CMP_XZR_XZR,
    asm,
    struct,
    _rd32,
)


def _u32(insn):
    return struct.unpack("<I", asm(insn))[0]


def _derive_mask_and_value(insns):
    vals = [_u32(i) for i in insns]
    mask = 0xFFFFFFFF
    for v in vals[1:]:
        mask &= ~(vals[0] ^ v)
    value = vals[0] & mask
    return mask, value


def _field_mask(total_bits=32, variable_fields=()):
    mask = (1 << total_bits) - 1
    for start, width in variable_fields:
        mask &= ~(((1 << width) - 1) << start)
    return mask & ((1 << total_bits) - 1)


class KernelJBPatchTaskConversionMixin:
    _ALLOW_SLOW_FALLBACK = (
        os.environ.get("VPHONE_TASK_CONV_ALLOW_SLOW_FALLBACK", "").strip() == "1"
    )

    # Build all matcher constants from keystone-assembled instruction bytes.
    # No hardcoded opcode constants.
    _CMP_XN_X0_MASK, _CMP_XN_X0_VAL = _derive_mask_and_value(
        ("cmp x0, x0", "cmp x1, x0", "cmp x30, x0")
    )
    _CMP_XN_X1_MASK, _CMP_XN_X1_VAL = _derive_mask_and_value(
        ("cmp x0, x1", "cmp x1, x1", "cmp x30, x1")
    )
    _BEQ_MASK = _field_mask(variable_fields=((5, 19),))
    _BEQ_VAL = _u32("b.eq #0x100") & _BEQ_MASK
    _LDR_X_UNSIGNED_MASK = _field_mask(variable_fields=((0, 5), (5, 5), (10, 12)))
    _LDR_X_UNSIGNED_VAL = _u32("ldr x0, [x0]") & _LDR_X_UNSIGNED_MASK
    _ADRP_MASK = 0x9F000000
    _ADRP_VAL = 0x90000000
    _BL_MASK = 0xFC000000
    _BL_VAL = 0x94000000
    _CBZ_W_MASK = 0x7F000000
    _CBZ_W_VAL = 0x34000000
    _CBNZ_W_VAL = 0x35000000
    _MOV_X19_X0 = _u32("mov x19, x0")
    _MOV_X0_X1 = _u32("mov x0, x1")

    def patch_task_conversion_eval_internal(self):
        """Allow task conversion: cmp Xn,x0 -> cmp xzr,xzr at unique guard site."""
        self._log("\n[JB] task_conversion_eval_internal: cmp xzr,xzr")

        ks, ke = self.kern_text
        candidates = self._collect_candidates_fast(ks, ke)
        # Fail-closed by default. Slow fallback can be explicitly enabled for
        # manual triage on unknown kernels.
        if len(candidates) != 1 and self._ALLOW_SLOW_FALLBACK:
            self._log(
                "  [!] fast matcher non-unique, trying slow fallback "
                "(VPHONE_TASK_CONV_ALLOW_SLOW_FALLBACK=1)"
            )
            candidates = self._collect_candidates_slow(ks, ke)

        if len(candidates) != 1:
            msg = (
                "  [-] expected 1 task-conversion guard site, found "
                f"{len(candidates)}"
            )
            if not self._ALLOW_SLOW_FALLBACK:
                msg += " (slow fallback disabled)"
            self._log(msg)
            return False

        self.emit(
            candidates[0], CMP_XZR_XZR, "cmp xzr,xzr [_task_conversion_eval_internal]"
        )
        return True

    @staticmethod
    def _decode_b_cond_target(off, insn):
        imm19 = (insn >> 5) & 0x7FFFF
        if imm19 & (1 << 18):
            imm19 -= 1 << 19
        return off + imm19 * 4

    def _is_candidate_context_safe(self, off, cmp_reg):
        # Require ADRP + LDR preamble for the same register.
        p2 = _rd32(self.raw, off - 8)
        if (p2 & self._ADRP_MASK) != self._ADRP_VAL:
            return False
        if (p2 & 0x1F) != cmp_reg:
            return False

        # Require the known post-compare sequence shape.
        if _rd32(self.raw, off + 16) != self._MOV_X19_X0:
            return False
        if _rd32(self.raw, off + 20) != self._MOV_X0_X1:
            return False

        i6 = _rd32(self.raw, off + 24)
        if (i6 & self._BL_MASK) != self._BL_VAL:
            return False

        i7 = _rd32(self.raw, off + 28)
        op = i7 & self._CBZ_W_MASK
        if op not in (self._CBZ_W_VAL, self._CBNZ_W_VAL):
            return False
        if (i7 & 0x1F) != 0:  # require w0
            return False

        # Both b.eq targets must be forward and nearby in the same routine.
        t1 = self._decode_b_cond_target(off + 4, _rd32(self.raw, off + 4))
        t2 = self._decode_b_cond_target(off + 12, _rd32(self.raw, off + 12))
        if t1 <= off or t2 <= off:
            return False
        if (t1 - off) > 0x200 or (t2 - off) > 0x200:
            return False
        return True

    def _collect_candidates_fast(self, start, end):
        cache = getattr(self, "_jb_scan_cache", None)
        key = ("task_conversion_fast", start, end)
        if cache is not None:
            cached = cache.get(key)
            if cached is not None:
                return cached

        out = []
        for off in range(start + 8, end - 28, 4):
            i0 = _rd32(self.raw, off)
            if (i0 & self._CMP_XN_X0_MASK) != self._CMP_XN_X0_VAL:
                continue
            cmp_reg = (i0 >> 5) & 0x1F

            p = _rd32(self.raw, off - 4)
            if (p & self._LDR_X_UNSIGNED_MASK) != self._LDR_X_UNSIGNED_VAL:
                continue
            p_rt = p & 0x1F
            p_rn = (p >> 5) & 0x1F
            if p_rt != cmp_reg or p_rn != cmp_reg:
                continue

            i1 = _rd32(self.raw, off + 4)
            if (i1 & self._BEQ_MASK) != self._BEQ_VAL:
                continue

            i2 = _rd32(self.raw, off + 8)
            if (i2 & self._CMP_XN_X1_MASK) != self._CMP_XN_X1_VAL:
                continue
            if ((i2 >> 5) & 0x1F) != cmp_reg:
                continue

            i3 = _rd32(self.raw, off + 12)
            if (i3 & self._BEQ_MASK) != self._BEQ_VAL:
                continue

            if not self._is_candidate_context_safe(off, cmp_reg):
                continue

            out.append(off)
        if cache is not None:
            cache[key] = out
        return out

    def _collect_candidates_slow(self, start, end):
        cache = getattr(self, "_jb_scan_cache", None)
        key = ("task_conversion_slow", start, end)
        if cache is not None:
            cached = cache.get(key)
            if cached is not None:
                return cached

        out = []
        for off in range(start + 4, end - 12, 4):
            d0 = self._disas_at(off)
            if not d0:
                continue
            i0 = d0[0]
            if i0.mnemonic != "cmp" or len(i0.operands) < 2:
                continue
            a0, a1 = i0.operands[0], i0.operands[1]
            if not (a0.type == ARM64_OP_REG and a1.type == ARM64_OP_REG):
                continue
            if a1.reg != ARM64_REG_X0:
                continue
            cmp_reg = a0.reg

            dp = self._disas_at(off - 4)
            d1 = self._disas_at(off + 4)
            d2 = self._disas_at(off + 8)
            d3 = self._disas_at(off + 12)
            if not dp or not d1 or not d2 or not d3:
                continue
            p = dp[0]
            i1, i2, i3 = d1[0], d2[0], d3[0]

            if p.mnemonic != "ldr" or len(p.operands) < 2:
                continue
            p0, p1 = p.operands[0], p.operands[1]
            if p0.type != ARM64_OP_REG or p0.reg != cmp_reg:
                continue
            if p1.type != ARM64_OP_MEM:
                continue
            if p1.mem.base != cmp_reg:
                continue

            if i1.mnemonic != "b.eq":
                continue
            if i2.mnemonic != "cmp" or len(i2.operands) < 2:
                continue
            j0, j1 = i2.operands[0], i2.operands[1]
            if not (j0.type == ARM64_OP_REG and j1.type == ARM64_OP_REG):
                continue
            if not (j0.reg == cmp_reg and j1.reg == ARM64_REG_X1):
                continue
            if i3.mnemonic != "b.eq":
                continue

            out.append(off)
        if cache is not None:
            cache[key] = out
        return out

File: sources/FirmwarePatcher/Kernel/JBPatches/KernelJBPatchTaskConversion.swift
// KernelJBPatchTaskConversion.swift — JB kernel patch: Task conversion eval bypass
//
// Historical note: derived from the legacy Python firmware patcher during the Swift migration.
//
// Strategy (fast raw scanner):
//   Locate the unique guard site in _task_conversion_eval_internal that performs:
//     ADRP Xn, <global>       ; [off - 8]  — loads global task-conversion table
//     LDR  Xn, [Xn]           ; [off - 4]  — dereferences it
//     CMP  Xn, X0             ; [off + 0]  — compare task pointer against X0
//     B.EQ <skip1>            ; [off + 4]
//     CMP  Xn, X1             ; [off + 8]  — compare against X1
//     B.EQ <skip2>            ; [off + 12]
//     MOV  X19, X0            ; [off + 16]
//     MOV  X0, X1             ; [off + 20]
//     BL   <callee>           ; [off + 24]
//     CBZ/CBNZ W0, ...        ; [off + 28]
//   Patch: replace CMP Xn, X0 with CMP XZR, XZR so the equality check always passes.

import Foundation

extension KernelJBPatcher {
    /// Task conversion eval bypass: patch the guard CMP to always be equal.
    @discardableResult
    func patchTaskConversionEvalInternal() -> Bool {
        log("\n[JB] task_conversion_eval_internal: cmp xzr,xzr")

        guard let range = kernTextRange ?? codeRanges.first.map({ ($0.start, $0.end) }) else {
            return false
        }
        let (ks, ke) = range

        let candidates = collectTaskConversionCandidates(start: ks, end: ke)

        guard candidates.count == 1 else {
            log("  [-] expected 1 task-conversion guard site, found \(candidates.count)")
            return false
        }

        let site = candidates[0]
        let va = fileOffsetToVA(site)
        emit(site, ARM64.cmpXzrXzr,
             patchID: "task_conversion_eval",
             virtualAddress: va,
             description: "cmp xzr,xzr [_task_conversion_eval_internal]")
        return true
    }

    // MARK: - Private scanner

    private func collectTaskConversionCandidates(start: Int, end: Int) -> [Int] {
        // Derived masks — no hardcoded opcode bytes:
        // CMP Xn, X0 = SUBS XZR, Xn, X0  → bits [31:21]=1110_1011_000, [20:16]=X0=00000,
        //   [15:10]=000000, [9:5]=Rn, [4:0]=11111(XZR)
        // Mask covers the fixed opcode and X0 operand; leaves Rn free.
        let cmpXnX0Mask: UInt32 = 0xFFFF_FC1F
        let cmpXnX0Val: UInt32 = 0xEB00_001F // cmp Xn, X0 — Rn wildcard

        // CMP Xn, X1 = SUBS XZR, Xn, X1  → Rm=X1=00001
        let cmpXnX1Mask: UInt32 = 0xFFFF_FC1F
        let cmpXnX1Val: UInt32 = 0xEB01_001F // cmp Xn, X1 — Rn wildcard

        // B.EQ #offset → bits[31:24]=0101_0100, bit[4]=0, bits[3:0]=0000 (EQ cond)
        let beqMask: UInt32 = 0xFF00_001F
        let beqVal: UInt32 = 0x5400_0000 // b.eq with any imm19

        // LDR Xd, [Xn] (unsigned offset, size=3):
        // bits [31:22] fixed = 0xF94 (size=11, V=0, opc=01, class=01);
        // bits [21:10] = imm12, bits [9:5] = Rn, bits [4:0] = Rt — all variable.
        let ldrXUnsignedMask: UInt32 = 0xFFC0_0000 // leaves imm12, Rn, Rt free
        let ldrXUnsignedVal: UInt32 = 0xF940_0000

        // ADRP: bit[31]=1, bits[28:24]=10000
        let adrpMask: UInt32 = 0x9F00_0000
        let adrpVal: UInt32 = 0x9000_0000

        // MOV X19, X0 = ORR X19, XZR, X0
        let movX19X0: UInt32 = 0xAA00_03F3
        // MOV X0, X1  = ORR X0, XZR, X1
        let movX0X1: UInt32 = 0xAA01_03E0

        // BL mask
        let blMask: UInt32 = 0xFC00_0000
        let blVal: UInt32 = 0x9400_0000

        // CBZ/CBNZ W (32-bit): bits[31]=0, bits[30:25]=011010 / 011011
        let cbzWMask: UInt32 = 0x7F00_0000
        let cbzWVal: UInt32 = 0x3400_0000 // CBZ  W
        let cbnzWVal: UInt32 = 0x3500_0000 // CBNZ W

        var out: [Int] = []

        var off = start + 8
        while off < end - 28 {
            defer { off += 4 }

            // [off]: CMP Xn, X0
            let i0 = buffer.readU32(at: off)
            guard (i0 & cmpXnX0Mask) == cmpXnX0Val else { continue }
            let cmpRn = (i0 >> 5) & 0x1F // the register being compared

            // [off - 4]: LDR Xn, [Xn] (load into cmpRn from cmpRn)
            let prev = buffer.readU32(at: off - 4)
            guard (prev & ldrXUnsignedMask) == ldrXUnsignedVal else { continue }
            let pRt = prev & 0x1F
            let pRn = (prev >> 5) & 0x1F
            guard pRt == cmpRn, pRn == cmpRn else { continue }

            // [off + 4]: B.EQ
            let i1 = buffer.readU32(at: off + 4)
            guard (i1 & beqMask) == beqVal else { continue }

            // [off + 8]: CMP Xn, X1 (same register)
            let i2 = buffer.readU32(at: off + 8)
            guard (i2 & cmpXnX1Mask) == cmpXnX1Val else { continue }
            guard ((i2 >> 5) & 0x1F) == cmpRn else { continue }

            // [off + 12]: B.EQ
            let i3 = buffer.readU32(at: off + 12)
            guard (i3 & beqMask) == beqVal else { continue }

            // Context safety: ADRP at [off - 8] for same register
            let p2 = buffer.readU32(at: off - 8)
            guard (p2 & adrpMask) == adrpVal else { continue }
            guard (p2 & 0x1F) == cmpRn else { continue }

            // [off + 16]: MOV X19, X0
            guard buffer.readU32(at: off + 16) == movX19X0 else { continue }

            // [off + 20]: MOV X0, X1
            guard buffer.readU32(at: off + 20) == movX0X1 else { continue }

            // [off + 24]: BL
            let i6 = buffer.readU32(at: off + 24)
            guard (i6 & blMask) == blVal else { continue }

            // [off + 28]: CBZ or CBNZ W0
            let i7 = buffer.readU32(at: off + 28)
            let op7 = i7 & cbzWMask
            guard op7 == cbzWVal || op7 == cbnzWVal else { continue }
            guard (i7 & 0x1F) == 0 else { continue } // must be W0

            // B.EQ targets must be forward and nearby (within same function)
            guard let t1 = decodeBEQTarget(insn: i1, at: off + 4) else { continue }
            guard let t2 = decodeBEQTarget(insn: i3, at: off + 12) else { continue }
            guard t1 > off, t2 > off else { continue }
            guard (t1 - off) <= 0x200, (t2 - off) <= 0x200 else { continue }

            out.append(off)
        }
        return out
    }

    /// Decode a B.cond target offset from an instruction at `pc`.
    private func decodeBEQTarget(insn: UInt32, at pc: Int) -> Int? {
        // B.cond: bits[31:24] = 0x54, bits[23:5] = imm19, bits[4] = 0, bits[3:0] = cond
        guard (insn & 0xFF00_001E) == 0x5400_0000 else { return nil }
        let imm19 = (insn >> 5) & 0x7FFFF
        // Sign-extend 19 bits
        let signedImm = Int32(bitPattern: imm19 << 13) >> 13
        return pc + Int(signedImm) * 4
    }
}
